### PR TITLE
Instrument `mod_pubsub`

### PR DESCRIPTION
**Pull request by gustawlippa**
This PR instruments `mod_pubsub`. Luckily, it war fairly straightforward.

I wasn't sure how to represent the event names without creating new atoms, which is why there is a new "table" in pubsub code, but I think it's OK, because: 1. It follows the style of the module (https://github.com/esl/MongooseIM/pull/4297/files#diff-8f3c86211cc3400d551754d19d3dbb12cecce86b38fe9f6ca2a7f2a2c87f5f05L1366 for example) 2. This way should be fast.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/221845) / elasticsearch_and_cassandra_mnesia / 3c2dda50ef370bcb8e4af60624c5068e16faec8b
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4297/221845/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4297/221845/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40f04a239a89dc.2024-06-11_10.32.06/index.html)
**OK: 456** / **Failed: 1** / **User-skipped: 41** / Auto-skipped: 0 
<details><summary>mam_SUITE:cassandra_muc_all:muc_prefs_cases:muc_prefs_set_cdata_request</summary>


```erlang
{error,
  {function_clause,
    [{mam_helper,parse_children_prefs_iq,
       [{xmlel,<<"error">>,
          [{<<"code">>,<<"501">>},{<<"type">>,<<"cancel">>}],
          [{xmlel,<<"feature-not-implemented">>,
             [{<<"xmlns">>,
             <<"urn:ietf:params:xml:ns:xmpp-stanzas">>}],
             []},
           {xmlel,<<"text">>,
             [{<<"xmlns">>,
             <<"urn:ietf:params:xml:ns:xmpp-stanzas">>}],
             [{xmlcdata,<<"From mod_mam_muc">>}]}]},
        {prefs_result_iq,undefined,[],[]}],
       []},
     {lists,foldl,3,[{file,"lists.erl"},{line,1594}]},
     {mam_SUITE,'-muc_prefs_set_cdata_request/1-fun-0-',2,
       [{file,"/home/circleci/project/big_tests/tests/mam_SUITE.erl"},
        {line,3457}]},
     {escalus_story,story,4,
       [{file,
          "/home/circleci/project/big_tests/_build/default/lib/escalus/src/escalus_story.erl"},
        {line,72}]},
     {muc_helper,story_with_room,4,
       [{file,"/home/circleci/project/big_tests/tests/muc_helper.erl"},
        {line,318}]},
     {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1793}]},
     {test_server,run_test_case_eval1,6,
       [{file,"test_server.erl"},{line,1302}]},
     {test_server,run_test_case_eval,9,
       [{file,"test_server.erl"},{line,1234}]}]}}
```

[Report log](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4297/221845/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40f04a239a89dc.2024-06-11_10.32.06/big_tests.tests.mam_SUITE.logs/run.2024-06-11_10.32.08/mam_suite.muc_prefs_set_cdata_request.10115.html)
</details>

File: big_tests/tests/pubsub_SUITE.erl [hunk 1]
```diff
@@ -352,7 +354,17 @@ subscribe_unsubscribe_test(Config) ->
...
               pubsub_tools:unsubscribe(Bob, Node, [{jid_type, bare}]),
 
-              pubsub_tools:delete_node(Alice, Node, [])
+              pubsub_tools:delete_node(Alice, Node, []),
+
+              BobJid = escalus_utils:get_jid(Bob),
+              instrument_helper:assert(mod_pubsub_set_subscribe, #{host_type => domain()},
+                                       fun(#{count := 1, jid := From, time := T}) ->
+                                           T >= 0 andalso BobJid =:= jid:to_binary(From)
+                                       end),
```
Comment (chrzaszcz) [inline big_tests/tests/pubsub_SUITE.erl:363]: How about a helper?

```
assert_event(EventName, Client) ->
   ClientJid = escalus_utils:get_jid(Client),
   instrument_helper:assert(EventName, #{host_type => domain()},
                            fun(#{count := 1, jid := From, time := T}) ->
                                T >= 0 andalso BobJid =:= jid:to_binary(From)
                            end),
```

File: src/pubsub/mod_pubsub.erl [hunk 2]
```diff
@@ -1313,18 +1312,16 @@ iq_pubsub(Host, ServerHost, From, IQType, #xmlel{children = SubEls} = QueryEl,
...
-            report_iq_action_metrics_after_return(ServerHost, Result, Time, IQType, Name),
-            Result;
+            mongoose_instrument:span(event_name(IQType, Name), #{host_type => ServerHost}, fun iq_pubsub_action/6,
+                                     [IQType, Name, Host, Node, From,
+                                      #{server_host => ServerHost,
+                                        plugins => Plugins,
+                                        access => Access,
+                                        action_el => ActionEl,
+                                        query_el => QueryEl,
+                                        lang => Lang}],
```
Comment (chrzaszcz) [inline src/pubsub/mod_pubsub.erl:1323]: Maybe construct this map separately to simplify this big expression?

**Review by chrzaszcz — COMMENTED** [review]
Looks good. I only added minor nodes.

File: big_tests/tests/pubsub_SUITE.erl (hunk 1 above)
Comment (gustawlippa) [reply big_tests/tests/pubsub_SUITE.erl]: Good idea, thanks! I've also added an async version. There is one place which checks for errors using a different fun, but I left it as it is now.

File: src/pubsub/mod_pubsub.erl (hunk 2 above)
Comment (gustawlippa) [reply src/pubsub/mod_pubsub.erl]: I've moved this map in both places, please see if this is something that you've had in mind. I think it looks better now indeed.

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/222131) / elasticsearch_and_cassandra_mnesia / 47a89410b8ca001f153cc78b0cecb21ee13bd46f
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4297/222131/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4297/222131/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40274f7c62109c.2024-06-12_12.31.54/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

File: src/pubsub/mod_pubsub.erl [hunk 3]
```diff
@@ -1396,23 +1392,48 @@ iq_action_to_metric_name(<<"purge">>) -> purge;
...
+event_name(get, subscriptions) ->
+    mod_pubsub_get_subscriptions;
+event_name(set, subscriptions) ->
+    mod_pubsub_set_subscriptions;
+event_name(get, affiliations) ->
+    mod_pubsub_get_affiliations;
+event_name(set, affiliations) ->
+    mod_pubsub_set_affiliations.
+
+ip_action_result_to_measurements(_Time, {error, _}, From) ->
```
Comment (chrzaszcz) [inline src/pubsub/mod_pubsub.erl:1433]: Shouldn't it be `iq_action_...`?

**Review by chrzaszcz — APPROVED** [review]
Looks good to me. I just added one comment.

File: src/pubsub/mod_pubsub.erl (hunk 3 above)
Comment (gustawlippa) [reply src/pubsub/mod_pubsub.erl:1433]: Of course... I don't know how it happened, and then it slipped me. Thanks for noticing, I'll fix it right away 👍

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/222265) / elasticsearch_and_cassandra_mnesia / 6d25e6b0faebdb9bcc9727bba56517fedf9cbc56
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4297/222265/elasticsearch_and_cassandra_mnesia.26.2.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4297/222265/elasticsearch_and_cassandra_mnesia.26.2.2-amd64/big/ct_run.test%40a65532ff737a.2024-06-13_07.32.02/index.html)
**OK: 457** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Review by arcusfelis — APPROVED** [review]
looks ok

**Pull request merged**